examples/ini.rs

```diff
@@ -0,0 +1,292 @@
+//! A basic rather tolerant ini parser, which most certainly does not obey to the standard
+use dangerous::{Error, Expected, Reader};
+use std::io::{self, Read};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut input_data = Vec::new();
+    io::stdin().read_to_end(&mut input_data)?;
+    let input = dangerous::input(input_data.as_ref());
+    match input.read_all::<_, _, Expected>(read_ini) {
+        Ok(ini) => println!("{:#?}", ini),
+        Err(e) => eprintln!("{:#}", e),
+    };
+    Ok(())
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Pair<'a> {
+    name: &'a str,
+    value: &'a str,
+}
+
+#[derive(Default, Debug, PartialEq, Eq)]
+struct Document<'a> {
+    globals: Vec<Pair<'a>>,
+    sections: Vec<Section<'a>>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct Section<'a> {
+    name: &'a str,
+    properties: Vec<Pair<'a>>,
+}
+
+fn read_ini<'i, E>(r: &mut Reader<'i, E>) -> Result<Document<'i>, E>
+where
+    E: Error<'i>,
+{
+    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
+    if r.at_end() {
+        return Ok(Document::default());
+    }
+    let (globals, sections) = match r.peek_u8()? {
```

Comment:
I would personally compact this up more like this!
```rust
fn read_ini<'i, E>(r: &mut Reader<'i, E>) -> Result<Document<'i>, E>
where
    E: Error<'i>,
{
    skip_whitespace_or_comment(r, ConsumeTo::NextToken);
    let document = match r.peek_u8().ok() {
        None => Document::default(),
        Some(b'[') => Document {
            globals: vec![],
            sections: read_sections(r)?,
        },
        Some(_) => Document {
            globals: read_zero_or_more_properties_until_section(r)?,
            sections: read_sections(r)?,
        },
    };
    Ok(document)
}
```

Reply:
A valid use case for `Reader::peek_u8_opt(..) -> Option<u8>`

Reply:
I like it a lot, it neatly combines 'EOF' with peeking. It also looks like this is the only possible error, so no actual error gets ignored that way.

Applied!

Reply:
Regarding `peek_u8_opt()`, I think this would make clear that no unexpected error will be lost, which makes it clearer than `peek_u8().ok()` and would indeed be preferred here.